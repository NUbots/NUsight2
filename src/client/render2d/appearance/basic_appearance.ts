import { observable } from 'mobx'

type Fill = { color: string, alpha: number }
type Stroke = { color: string, alpha: number, width: number }

export type BasicAppearanceOpts = {
  fill?: Partial<Fill>,
  stroke?: Partial<Stroke>
}

export class BasicAppearance {
  @observable fill?: Fill
  @observable stroke?: Stroke

  constructor({ fill, stroke }: { fill?: Fill, stroke?: Stroke }) {
    this.fill = fill
    this.stroke = stroke
  }

  static of({ fill, stroke }: BasicAppearanceOpts = {}): BasicAppearance {
    return new BasicAppearance({
      fill: fill && { color: '#000000', alpha: 1, ...fill },
      stroke: stroke && { color: '#000000', alpha: 1, width: 1, ...stroke },
    })
  }
}
