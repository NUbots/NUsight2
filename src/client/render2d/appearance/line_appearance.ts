import { observable } from 'mobx'


export type LineAppearanceOpts = {
  stroke?: {
    color: string,
    alpha?: number,
    width?: number,
    cap?: 'butt' | 'round' | 'square'
    dashOffset?: number
    join?: 'bevel' | 'round' | 'miter'
  }
}

export class LineAppearance {
  @observable stroke: {
    color: string,
    alpha: number,
    width: number,
    cap: 'butt' | 'round' | 'square'
    dashOffset: number
    join: 'bevel' | 'round' | 'miter'
  } = {
    color: '#000000',
    alpha: 1,
    width: 1,
    cap: 'butt',
    dashOffset: 0,
    join: 'miter',
  }

  constructor(opts: LineAppearanceOpts) {
    Object.assign(this.stroke, opts.stroke)
  }

  static of(opts: LineAppearanceOpts = {}): LineAppearance {
    return new LineAppearance(opts)
  }
}
