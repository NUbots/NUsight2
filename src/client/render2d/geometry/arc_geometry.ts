import { observable } from 'mobx'

import { Vector2 } from '../../math/vector2'

export class ArcGeometry {
  @observable origin: Vector2
  @observable radius: number
  @observable startAngle: number
  @observable endAngle: number
  @observable anticlockwise: boolean

  constructor(opts: ArcGeometry) {
    this.origin = opts.origin
    this.radius = opts.radius
    this.startAngle = opts.startAngle
    this.endAngle = opts.endAngle
    this.anticlockwise = opts.anticlockwise
  }

  static of({
              origin = Vector2.of(0, 0),
              radius = 1,
              startAngle = 0,
              endAngle = 2 * Math.PI,
              anticlockwise = false,
            }: Partial<ArcGeometry> = {}): ArcGeometry {
    return new ArcGeometry({
      origin,
      radius,
      startAngle,
      endAngle,
      anticlockwise,
    })
  }
}
