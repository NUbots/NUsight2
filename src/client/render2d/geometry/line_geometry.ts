import { observable } from 'mobx'

import { Vector2 } from '../../math/vector2'

export class LineGeometry {
  @observable origin: Vector2
  @observable target: Vector2

  constructor(opts: LineGeometry) {
    this.origin = opts.origin
    this.target = opts.target
  }

  static of({
              origin = Vector2.of(0, 0),
              target = Vector2.of(1, 1),
            }: Partial<LineGeometry> = {}): LineGeometry {
    return new LineGeometry({
      origin,
      target,
    })
  }
}
