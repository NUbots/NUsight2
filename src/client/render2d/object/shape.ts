import { observable } from 'mobx'

import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'

import { Geometry } from './geometry'

export type ShapeOpts<T extends Geometry> = {
  appearance: Appearance
  geometry: T
}

export class Shape<T extends Geometry> {
  @observable appearance: Appearance
  @observable geometry: T

  constructor(opts: ShapeOpts<T>) {
    this.appearance = opts.appearance
    this.geometry = opts.geometry
  }

  static of<T extends Geometry>(geometry: T, appearance: Appearance = BasicAppearance.of()) {
    return new Shape<T>({
      appearance,
      geometry,
    })
  }
}
