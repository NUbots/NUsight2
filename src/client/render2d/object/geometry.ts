import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'

export type Geometry =
  ArcGeometry
  | ArrowGeometry
  | CircleGeometry
  | LineGeometry
  | PathGeometry
  | PolygonGeometry
  | TextGeometry
