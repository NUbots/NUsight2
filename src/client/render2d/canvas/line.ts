import { LineGeometry } from '../geometry/line_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export function renderLine(ctx: CanvasRenderingContext2D, shape: Shape<LineGeometry>): void {

  const { origin, target } = shape.geometry

  ctx.beginPath()
  ctx.moveTo(origin.x, origin.y)
  ctx.lineTo(target.x, target.y)

  applyAppearance(ctx, shape.appearance)

  ctx.stroke()
}
