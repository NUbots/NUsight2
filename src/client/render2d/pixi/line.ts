import { createTransformer } from 'mobx-utils'
import { Graphics } from 'pixi.js'

import { LineGeometry } from '../geometry/line_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export const renderLine = createTransformer((shape: Shape<LineGeometry>): Graphics => {

  const { origin, target } = shape.geometry

  const g = new Graphics()

  applyAppearance(g, shape.appearance, g => {
    g.moveTo(origin.x, origin.y)
    g.lineTo(target.x, target.y)
  })

  return g
})
