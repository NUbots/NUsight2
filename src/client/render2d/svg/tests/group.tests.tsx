import * as React from 'react'

import { Transform } from '../../../math/transform'
import { BasicAppearance } from '../../appearance/basic_appearance'
import { LineGeometry } from '../../geometry/line_geometry'
import { Shape } from '../../object/shape'
import { Line } from '../line'

describe('LineSVGRenderer', () => {
  it('renders', () => {
    const model = Shape.of(LineGeometry.of(), BasicAppearance.of())
    const world = Transform.of()
    expect(<Line model={model} world={world}/>).toMatchSnapshot()
  })
})
