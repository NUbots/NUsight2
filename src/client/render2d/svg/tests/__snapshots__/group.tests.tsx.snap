// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineSVGRenderer renders 1`] = `
<_class
  model={
    Shape {
      "appearance": BasicAppearance {
        "fill": undefined,
        "stroke": undefined,
      },
      "geometry": LineGeometry {
        "origin": Vector2 {
          "x": 0,
          "y": 0,
        },
        "target": Vector2 {
          "x": 1,
          "y": 1,
        },
      },
    }
  }
  world={
    Transform {
      "anticlockwise": true,
      "rotate": 0,
      "scale": Object {
        "x": 1,
        "y": 1,
      },
      "translate": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
/>
`;
